Refactor DBLayer to use in-memory Checkpoints

HeinrichApfelmus opened the pull request:
### Issue number

ADP-1169

### Overview

Previous work in epic ADP-1043 introduced delta encodings, DBVars, and an embedding of the wallet state and its delta encodings into a database table. It's time to integrate these tools with the wallet code. To facilitate code review, the integration proceeds in a sequence of refactorings that do not change functionality and pass all unit tests.

In this step, we introduce a type Checkpoints a which stores all checkpoints of the wallet state in memory. The current ad-hoc cache of the latest checkpoints is replaced by a DBVar that stores these checkpoints instead.

### Details

* Keep the `Checkpoints` type stupid and simple for now. It will be replaced later.
* This pull request will (temporarily) increase the memory requirements for the wallet, as the collection of checkpoints is kept in-memory. However, memory usage will go back to normal (and better) once delta encodings are integrated.

### Comments

Merge PR #2841 before this one, because this pull request is based on the branch of the former.

Commit:
Introduce `DBVar (Checkpoints (W.Wallet s))`

* Keep the `Checkpoints` type stupid and simple for now. It will be replaced later.
* This replaces the cache of the latest checkpoint with a cache of all checkpoints. This will increase memory requirements for now, but this will go back to normal (and better) once delta encodings are integrated.

Commit:
Introduce `Slot` type to distinguish genesis from (SlotNo 0)

This commit partially addresses the issue where the DB layer historically did not distinguish between the genesis point and the block with SlotNo 0, which comes directly after genesis.

* The types of `rollbackTo` and `listCheckpoints` can now distinguish these points.
* For reasons of correctness, we need to use `Slot` in the `Checkpoints` type.
* However, the DB file format still uses slot numbers only. But as we want to revamp the format anyway, the plan is to keep it as it is for now, and remove the issue with the revamp.

paweljakubas reviewed (commented):
I run through logic and it seems to proceed as I would expect. If all tests pass it is giving crucial confidence that replacing SlotNo with SlotPoint and BlockHeader with ChainPoint is not doing any harm. What about db migrations? No way they are going to be affected? Maybe it would be good if @piotr-iohk plays with migrations using this branch.

I will give another look, even more detailed and style related, tomorrow (until now I was mainly occupied with how it works and why)

Commit:
Change `BlockHeader` to use Maybe for parent hash

The genesis block has no parent block, and this fact is best represented by `Nothing` instead of a magic value `hashOfNoParent`.

The magic value is still used for serialization and deserialization in the database.

paweljakubas reviewed (approved):
lgtm

rvl commented on lib/core/src/Cardano/Wallet/DB.hs:
```diff
@@ -301,12 +302,13 @@ data DBLayer m s k = forall stm. (MonadIO stm, MonadFail stm) => DBLayer
-        -> SlotNo
-        -> ExceptT ErrNoSuchWallet stm BlockHeader
-        -- ^ Drops all checkpoints and transaction data after the given slot.
+        -> Slot
```
Given that we don't require the database to be rolled back further than a few thousand blocks, a value of `Origin` should never be passed for this argument of `rollbackTo`.

HeinrichApfelmus replied:
I agree that this will not happen in practice on testnet and mainnet. However, the ouroboros-network library does represent points on the blockchain with an explicit origin, and a rollback to genesis can happen in principle (and does happen on local test clusters). By using a representation that is more closely aligned with ouroboros-network (i.e. `Slot` instead of `SlotNo`), the code actually becomes simpler, as we can remove conversion functions such as

```haskell
    pseudoPointSlot :: ChainPoint -> SlotNo
    pseudoPointSlot ChainPointAtGenesis = W.SlotNo 0
    pseudoPointSlot (ChainPoint slot _) = slot
```

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -2050,27 +2048,191 @@ genesisParametersFromEntity (Wallet _ _ _ _ _ hash startTime) =
+
+'SlotNo' cannot represent the genesis point.
+
+Historical hack. The DB layer can't represent 'Origin' in the database,
+instead we have mapped it to 'SlotNo 0', which is wrong.
```
It shouldn't matter; we don't need rollback to genesis.

rvl reviewed (approved):
Great :+1:
